### PR TITLE
init: frontmatter support

### PR DESCRIPTION
**Pull request by hsjobeki**
### Why frontmatter is needed (sometimes)

Sometimes it is desireable to extend the content with meta information. Frontmatter is the de-facto standard of adding document specific meta tags to markdown. [1](https://docs.github.com/en/contributing/writing-for-github-docs/using-yaml-frontmatter) [2](https://jekyllrb.com/docs/front-matter/) [3](https://starlight.astro.build/reference/frontmatter/)

This would allow us to keep track of references between documents that are closely related to nix code or externalized doc-comments that are no longer tracked.

This PR also includes a detailed [design document](/doc/frontmatter.md). 

i.e.,

- Enriching documentation generation.
- Maintaining References.
- Metadata inclusion.
- Allow better Handling of edge cases.

### Example

```nix
{
/** 
 ---
 import: ./path.md
 ---
*/
foo = x: x;
}
```

## TODO

- [ ] Unit tests
- [ ] Bump version
- [ ] Manual testing with nixpkgs/nixos manuals

### CODE REVIEW

File: doc/frontmatter.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+## Keywords
+
+### Import
+
+Rules:
+
+1. The `import` keyword can be used to use content from another file INSTEAD of the doc-comments content.
+
+2. The value `file` must be given as an absolute path (relative to git root) or relative to the current file.
```
Comment (infinisil) [inline doc/frontmatter.md:54]: Should absolute paths be allowed?
- (+) When the docs are entirely elsewhere, e.g. `doc/manual/..`, a relative path would have to be `../../..`, very ugly
  - (-) If only relative paths are allowed, encourages moving docs closer to the source
    
    Makes changing documentation easier
    - For the Nixpkgs manual build, needs to adjust which files are included in the derivation source
  - (-) With only relative paths, it's more similar to NixOS module docs
- (-) Makes it very confusing with where absolute paths are relative to (build root, git root, `.nix` location, etc.)
- (-) We can still allow absolute paths later on if necessary
- (-) Relies on a Git repository and git installed
- (-) It's unclear where absolute paths are rooted
  - (+) Could use a syntax like `$GIT_ROOt/foo/bar`
    - (-) Not a fan of more custom syntax

Decision: Don't allow for now

This can be added to the design document (I recommend `<details> </details>`)

File: doc/frontmatter.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+- Metadata inclusion.
+- Allow better Handling of edge cases.
+
+## Detailed design
+
+Fields (from Keywords list) can be defined in frontmatter.
+
+Frontmatter is defined using key-value pairs, encapsulated within triple-dashed lines (---).
+
+While there is no strict specification for frontmatter formats, YAML is commonly preferred. Although JSON could also be used alternatively.
```
Comment (DanielSidhion) [inline doc/frontmatter.md:24]:
```suggestion
While there is no strict specification for frontmatter formats, YAML is commonly preferred.
Although JSON could also be used alternatively.
```

File: doc/frontmatter.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+`path.md`
+```md
+some nice docs!
+```
+
+Rendering this would behave as if the content where actually placed in the doc-comment itself.
+
+Placing frontmatter inside the imported file will be ignored. (No nested directives)
+Since handling recursive or nested imports adds too much complexity for little or no benefit.
```
Comment (infinisil) [inline doc/frontmatter.md:77]: Recommended rewording to: nixdoc does not process any content inside the file, using it as-is

File: doc/frontmatter.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+foo = x: x;
+}
+```
+
+`path.md`
+```md
+some nice docs!
+```
+
+Rendering this would behave as if the content where actually placed in the doc-comment itself.
```
Comment (infinisil) [inline doc/frontmatter.md:74]: With the switch to `doc-comment` name, makes more sense to formulate this as "defer to" instead of "pull"

File: doc/frontmatter.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+Placing frontmatter inside the imported file will be ignored. (No nested directives)
+Since handling recursive or nested imports adds too much complexity for little or no benefit.
+
+> Note: Absolute path imports are relative to the repository root. They only work inside `git` repositories and require having the `git` binary in PATH.
+> This is most commonly used in large repositories or when the documentation files are not placed alongside the .nix files.
+
+## Extensibility
+
+The initial set of keywords is intentionally minimalistic,
+focusing on immediate and broadly applicable needs.
```
Comment (DanielSidhion) [inline doc/frontmatter.md:85]:
```suggestion
The initial set of keywords is intentionally minimalistic, focusing on immediate and broadly applicable needs.
```

File: src/frontmatter.rs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+    match YamlFrontMatter::parse::<Matter>(markdown.unwrap()) {
+        Ok(document) => {
+            let metadata = document.metadata.content;
+
+            let abs_import = metadata.get("import").map(|field| {
+                let import_val = field
+                    .as_str()
+                    .expect("Frontmatter: import field must be a string");
+                match PathBuf::from(import_val).is_relative() {
```
Comment (infinisil) [inline src/frontmatter.rs:40]: I can recommend https://docs.rs/relative-path to deal with relative paths

File: src/frontmatter.rs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+                    ]),
+                }
+            });
+
+            abs_import.map(|path| {
+                fs::read_to_string(&path)
+                    .expect(format!("Could not read file: {:?}", &path).as_str())
+            })
+        }
+        Err(_) => None,
```
Comment (infinisil) [inline src/frontmatter.rs:59]: A bit more error handling would be good, though it looks like the library is a bit of a pain

File: src/main.rs [hunk 8]
```diff
@@ -99,12 +101,19 @@ enum DocItemOrLegacy {
...
+    shift_headings_by: Option<usize>,
+    file: &Path,
+) -> Option<String> {
     let doc_comment = get_expr_docs(node);
 
-    doc_comment.map(|doc_comment| {
+    // Doc comments can import external file via "import" in frontmatter
+    let content = get_imported_content(file, doc_comment.as_ref()).or(doc_comment);
+
+    content.map(|inner| {
```
Comment (infinisil) [inline src/main.rs:114]:
```suggestion
    content.map(|inner| {
        // Doc comments can import external file via "import" in frontmatter
        let inner_new = get_imported_content(file, inner).or(inner);
```

This way you can avoid the `Option` in the function

**Review by infinisil — CHANGES_REQUESTED** [review]
Reviewed together in the docs team meeting!

File: src/frontmatter.rs (hunk 7 above)
Comment (hsjobeki) [reply src/frontmatter.rs:59]: I use gray_matter now. Error handling is still a pain, but figured out how it works.

**Comment by hsjobeki** [comment]
@infinisil I think this PR is ready from my side. If you could give it a quick revise, if all issues are addressed. 

Just tested it against nixpkgs with some lib functions. works really nice, didn't notice any problems or issues with the other tooling whatsoever.

Note: I switched to another frontmatter parser library because the error handling wasn't expressive enough.

File: README.md [hunk 9]
```diff
@@ -60,10 +63,39 @@ The following is an example of markdown documentation for new and current users
 
 > Note: Within nixpkgs the convention of using [definition-lists](https://www.markdownguide.org/extended-syntax/#definition-lists) for documenting arguments has been established.
 
+## Example: using frontmatter
```
Comment (DanielSidhion) [inline README.md:66]:
```suggestion
#### Example: using frontmatter
```

File: doc/frontmatter.md [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    tags:
+      - foo
+      - bar
+    ---
+  */
+  foo = x: x;
+}
+```
+
+In this example `key` is `tags` and `value` is `List("foo", "bar")`.
```
Comment (DanielSidhion) [inline doc/frontmatter.md:46]: Very minor suggestion, but since we don't support a `tags` key, let's just make it generic. Bonus is that we can just reuse the `key` and `value` terms used above without having to further clarify ("`key` is `tags` and ...")

~~~suggestion
    key: value
    ---
  */
  foo = x: x;
}
```
~~~

File: doc/frontmatter.md [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+See also: [yaml specification](https://yaml.org/spec/1.2.2/) for how to use YAML.
+
+## Keywords
+
+### `doc_location`
+
+Rules:
+
+1. The `doc_location` keyword can be used to use content from another file INSTEAD of the doc-comments content.
```
Comment (DanielSidhion) [inline doc/frontmatter.md:56]: During the meeting we discussed forbidding any extra doc-comment content entirely:

```suggestion
1. The `doc_location` keyword can be used to use content from another file and **forbid** any further content to be added to the existing doc-comment.
```

File: doc/frontmatter.md [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+2. The value must be given as a path relative to the current file.
+
+3. Using this directive does not process any content inside the file, using it as-is.
```
Comment (DanielSidhion) [inline doc/frontmatter.md:60]: This rewording makes it clearer which file you're talking about.

```suggestion
3. The file pointed to by `doc_location` will be used as-is, without any further processing done to it.
```

File: doc/frontmatter.md [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+  foo = x: x;
+}
+```
+
+`path.md`
+```md
+some nice docs!
+```
+
+In this example, the `doc_location` directive fetches content from `./path.md` treating it as if it were directly within the doc-comment.
```
Comment (DanielSidhion) [inline doc/frontmatter.md:78]: This rewording avoids anyone thinking they could possibly include `doc_location` and also add content in the doc-comment, and have both contents get merged.

```suggestion
In this example, the `doc_location` directive fetches content from `./path.md` and treats it as the actual doc-comment.
```

File: doc/frontmatter.md [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+This allows tracking the reference between the source position and the markdown file, in case of external documentation.
+
+#### Design decision: Relative vs Absolute paths
+
+This section justifies the previous decision why absolute paths are not allowed.
```
Comment (DanielSidhion) [inline doc/frontmatter.md:83]: You haven't said that absolute paths are not allowed anywhere until now. I suggest rewording:

```suggestion
This section explains the decision to only use relative paths and not support absolute paths for the file pointed to by `doc_location`.
```

File: doc/frontmatter.md [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+
+<details>
+<summary>Should absolute paths be allowed?</summary>
+
+- (+) When the docs are entirely elsewhere, e.g. `doc/manual/..`, a relative path would have to be `../../..`, very ugly
+  - (-) If only relative paths are allowed, encourages moving docs closer to the source,
+    Makes changing documentation easier.
```
Comment (DanielSidhion) [inline doc/frontmatter.md:90]:
```suggestion
  - (-) If only relative paths are allowed, encourages moving docs closer to the source, makes changing documentation easier.
```

File: doc/frontmatter.md [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+    - For the nix-build, adjustments which files are included in the derivation source may be needed.
```
Comment (DanielSidhion) [inline doc/frontmatter.md:91]:
```suggestion
    - For the nix-build, adjustments of which files are included in the derivation source may be needed.
```

File: doc/frontmatter.md [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+  - (-) With only relative paths, it's more similar to NixOS module docs
+- (-) Makes it very confusing with where absolute paths are relative to (build root, git root, `.nix` location, etc.)
+- (-) We can still allow absolute paths later on if necessary
+- (-) Relies on a Git repository and git installed
+- (-) It's unclear where absolute paths are rooted
+  - (+) Could use a syntax like `$GIT_ROOt/foo/bar`
+    - (-) Not a fan of more custom syntax
```
Comment (DanielSidhion) [inline doc/frontmatter.md:98]:
```suggestion
- (-) We can still allow absolute paths later on if necessary
- (-) Makes it very confusing where absolute paths are relative to (build root, git root, `.nix` location, etc.)
  - (+) Could use a syntax like `$GIT_ROOT/foo/bar`
    - (-) Relies on a Git repository and git installed
    - (-) Not a fan of more custom syntax
```

File: doc/frontmatter.md [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+
+**Decision**: Not supported by now.
+
+This outcome was discussed in the nix documentation team: https://discourse.nixos.org/t/2024-03-21-documentation-team-meeting-notes-114/41957.
```
Comment (DanielSidhion) [inline doc/frontmatter.md:102]:
```suggestion
This outcome was discussed in the nix documentation team meeting: https://discourse.nixos.org/t/2024-03-21-documentation-team-meeting-notes-114/41957.
```

File: doc/frontmatter.md [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+## Error handling
+
+Any issues encountered during the processing of frontmatter — be it syntax errors, invalid paths, or unsupported keywords—should result in clear, actionable error messages to the user.
+
+## Extensibility
+
+The initial set of keywords is intentionally minimalistic, focusing on immediate and broadly applicable needs.
+
+When extending this document we ask our contributors to be tooling agnostic, such that documentation wont't rely on any implementation details.
+This approach ensures that the documentation remains independent of specific implementation details. By adhering to this principle, we aim to create a resource that is universally applicable
```
Comment (DanielSidhion) [inline doc/frontmatter.md:115]:
```suggestion
This approach ensures that the documentation remains independent of specific implementation details. By adhering to this principle, we aim to create a resource that is universally applicable.
```

**Review by DanielSidhion — COMMENTED** [review]
Haven't looked at the code. My review is purely on the design document and README.

File: doc/frontmatter.md (hunk 11 above)
Comment (hsjobeki) [reply doc/frontmatter.md:56]: Ok. but this, should then also be checked if possible. Going to add an additional check to the implementation.

**Review by DanielSidhion — APPROVED** [review]
Approving based only on documentation content, I'm leaving the code review to @infinisil since I have 0 experience on nixdoc code.

**Comment by fricklerhandwerk** [comment]
Reviewed in the docs team meeting:
- We've discussed this quite intensely, because I have strong doubts this is not an anti-pattern in the long run
- The motivation as stated is a bit weak; the only concrete problems it would solve are:
  - Editing large doc strings has better tooling support in plain markdown files
  - Avoiding large doc strings in the source makes the code slightly easier to navigate
- On the other hand:
  - It adds a layer of indirection ('invisible behavior' you have to "kinda know it exists"), e.g. what exactly does `doc-location` mean?
  - You need to navigate and edit multiple files if the pattern is used
  - I say: When the pattern is used, it's proliferated over time, and I fear it will add friction below the threshold of pain. So it won't get reported to actually get mitigation.
- This is a trade-off, and there's no obviously correct solution
- Agreement: Postpone merging until we have a clear-cut use case that would be to painful to support without this feature

**Comment by hsjobeki** [comment]
> When the pattern is used, it's proliferated over time, and I fear it will add friction below the threshold of pain.

@fricklerhandwerk @infinisil.

This fear is justified, and I share it. Nevertheless, it would be very beneficial to provide this feature for individual edge cases with very long documentation. This should not be used extensively, and there are only a handful of places where I can imagine this.
For example, the "mkDerivation" function documentation is rather extensive, and you wouldn't want that in the source code.

So nixdoc is just a documentation tool that offers this feature. It is not an official tool or an opinionated tool. But nixpkgs should adhere to usage guidelines. e.g. to use the doc_location feature only for certain functions.

I am still demanding this feature and planning to add more features to "frontmatter" later on, which might be needed in my vision for improved documentation.

**Comment by hsjobeki** [comment]
@infinisil @fricklerhandwerk
Another place that i think may profit from this feature.

`evalModules`
https://github.com/NixOS/nixpkgs/blob/18bcb1ef6e5397826e4bfae8ae95f1f88bf59f4f/lib/modules.nix#L85
Documentation file > 100 lines
https://github.com/NixOS/nixpkgs/blob/master/doc/module-system/module-system.chapter.md

Moving documentation into the .nix source files seems not beneficial. 
We could add a limit (say 50 lines) to avoid this feature beeing overused.    

This feature can help in some places to restructure the content of documentation into pieces that are directly related to source (hence reference) and the other pieces (guide, tutorials, ... ) 

Therefore i propose to merge this.

**Comment by hsjobeki** [comment]
There are some edge cases where position tracking in nix is insufficient yet.

I have the following idea that some doc-comments may include the full path of the nix value they document which allows a more decoupled evolution of the doc-comments from the nix position tracking.

```
---
path = "pkgs.pythonPackage.untrackable"
---

Some content for an edge case, untrackable function
```
